File: GitHub_task/task.py
from abc import ABC, abstractmethod
from typing import List, Optional
import copy

class Printable(ABC):
    """Base abstract class for printable objects."""
    
    def print_me(self, os, prefix="", is_last=False, no_slash=False, is_root=False):
        """Base printing method for the tree structure display.
        Implement properly to display hierarchical structure."""
        # To be implemented
        pass
        
    @abstractmethod
    def clone(self):
        """Create a deep copy of this object."""
        pass

class BasicCollection(Printable):
    """Base class for collections of items."""
    def __init__(self):
        self.items = []
    
    def add(self, elem):
        self.items.append(elem)
        return self
    
    def find(self, name):
        for item in self.items:
            if isinstance(item, Computer) and item.name == name:
                return item
            if isinstance(item, BasicCollection):
                found = item.find(name)
                if found:
                    return found
        return None


class Component(Printable):
    """Base class for computer components."""
    def __init__(self, numeric_val=0):
        self.numeric_val = numeric_val
        
    def clone(self):
        return copy.deepcopy(self)

class Address(Printable):
    """Class representing a network address."""
    def __init__(self, addr):
        self.address = addr
    
    # To be implemented
    def print_me(self, prefix="", is_last=False, no_slash=False, is_root=False) -> str:
        symbol = "\\-" if is_last else "+-"
        return f"{prefix}{symbol}{self.address}\n"

    
    def clone(self):
        return Address(self.address)

class Computer(BasicCollection, Component):
    """Class representing a computer with addresses and components."""
    def __init__(self, name):
        super().__init__()
        self.name = name
        self.addresses = []
        self.components = []
    
    def add_address(self, addr):
        self.addresses.append(Address(addr))
        return self
    
    def add_component(self, comp):
        self.components.append(comp)
        return self
    
    def find(self, name):
        if self.name == name:
            return self
        for comp in self.components:
            if isinstance(comp, BasicCollection):
                found = comp.find(name)
                if found:
                    return found
        return None

    
    def print_me(self, prefix="", is_last=False, no_slash=False, is_root=False) -> str:
        symbol = "\\-" if is_last else "+-"
        result = f"{prefix}{symbol}Host: {self.name}\n"
        child_prefix = prefix + ("  " if is_last else "| ")
        total = self.addresses + self.components
        for i, item in enumerate(total):
            result += item.print_me(child_prefix, is_last=(i == len(total) - 1))
        return result

    def clone(self):
        new_comp = Computer(self.name)
        new_comp.addresses = [a.clone() for a in self.addresses]
        new_comp.components = [c.clone() for c in self.components]
        return new_comp

class Network(Printable):
    """Class representing a network of computers."""
    def __init__(self, name):
        self.name = name
        self.computers = []
    
    def add_computer(self, comp):
        self.computers.append(comp)
        return self
    
    def find_computer(self, name):
        for c in self.computers:
            found = c.find(name)
            if found:
                return found
        return None
    
    # Другие методы...
    def clone(self):
        new_net = Network(self.name)
        new_net.computers = [c.clone() for c in self.computers]
        return new_net
    
    def print_me(self, prefix="", is_last=False, no_slash=False, is_root=False) -> str:
        result = f"Network: {self.name}\n"
        for i, comp in enumerate(self.computers):
            result += comp.print_me("", is_last=(i == len(self.computers) - 1))
        return result

    def __str__(self):
        return self.print_me().strip()

class Disk(Component):
    """Disk component class with partitions."""
    # Определение типов дисков
    SSD = 0
    MAGNETIC = 1
    
    def __init__(self, storage_type, size):
        # Initialize properly
        super().__init__(size)
        self.storage_type = storage_type
        self.partitions = []
    
    def add_partition(self, size, name):
        self.partitions.append((size, name))
        return self
    
    def print_me(self, prefix="", is_last=False, no_slash=False, is_root=False) -> str:
        symbol = "\\-" if is_last else "+-"
        typ = "SSD" if self.storage_type == Disk.SSD else "HDD"
        result = f"{prefix}{symbol}{typ}, {self.numeric_val} GiB\n"
        child_prefix = prefix + ("  " if is_last else "| ")
        for i, part in enumerate(self.partitions):
            part_symbol = "\\-" if i == len(self.partitions) - 1 else "+-"
            result += f"{child_prefix}{part_symbol}[{i}]: {part[0]} GiB, {part[1]}\n"
        return result

    def clone(self):
        new_disk = Disk(self.storage_type, self.numeric_val)
        new_disk.partitions = copy.deepcopy(self.partitions)
        return new_disk
    
    def __str__(self):
        return self.print_me().strip()

class CPU(Component):
    """CPU component class."""
    def __init__(self, cores, mhz):
        super().__init__()
        self.cores = cores
        self.mhz = mhz
    
    def print_me(self, prefix="", is_last=False, no_slash=False, is_root=False) -> str:
        symbol = "\\-" if is_last else "+-"
        return f"{prefix}{symbol}CPU, {self.cores} cores @ {self.mhz}MHz\n"

    def clone(self):
        return CPU(self.cores, self.mhz)

class Memory(Component):
    """Memory component class."""
    def __init__(self, size):
        super().__init__(size)
        self.size = size
    
    def print_me(self, prefix="", is_last=False, no_slash=False, is_root=False) -> str:
        symbol = "\\-" if is_last else "+-"
        return f"{prefix}{symbol}Memory, {self.size} MiB\n"

    def clone(self):
        return Memory(self.size)

# Пример использования (может быть неполным или содержать ошибки)
def main():
    # Создание тестовой сети
    n = Network("MISIS network")
    
    # Добавляем первый сервер с одним CPU и памятью
    n.add_computer(
        Computer("server1.misis.ru")
        .add_address("192.168.1.1")
        .add_component(CPU(4, 2500))
        .add_component(Memory(16000))
    )
    
    # Добавляем второй сервер с CPU и HDD с разделами
    n.add_computer(
        Computer("server2.misis.ru")
        .add_address("10.0.0.1")
        .add_component(CPU(8, 3200))
        .add_component(
            Disk(Disk.MAGNETIC, 2000)
            .add_partition(500, "system")
            .add_partition(1500, "data")
        )
    )
    
    # Выводим сеть для проверки форматирования
    print("=== Созданная сеть ===")
    print(n)
    
    # Тест ожидаемого вывода
    expected_output = """Network: MISIS network
+-Host: server1.misis.ru
| +-192.168.1.1
| +-CPU, 4 cores @ 2500MHz
| \-Memory, 16000 MiB
\-Host: server2.misis.ru
  +-10.0.0.1
  +-CPU, 8 cores @ 3200MHz
  \-HDD, 2000 GiB
    +-[0]: 500 GiB, system
    \-[1]: 1500 GiB, data"""
    
    assert str(n) == expected_output, "Формат вывода не соответствует ожидаемому"
    print("✓ Тест формата вывода пройден")
    
    # Тестируем глубокое копирование
    print("\n=== Тестирование глубокого копирования ===")
    x = n.clone()
    
    # Тестируем поиск компьютера
    print("Поиск компьютера server2.misis.ru:")
    c = x.find_computer("server2.misis.ru")
    print(c)
    
    # Модифицируем найденный компьютер в копии
    print("\nДобавляем SSD к найденному компьютеру в копии:")
    c.add_component(
        Disk(Disk.SSD, 500)
        .add_partition(500, "fast_storage")
    )
    
    # Проверяем, что оригинал не изменился
    print("\n=== Модифицированная копия ===")
    print(x)
    print("\n=== Исходная сеть (должна остаться неизменной) ===")
    print(n)
    
    # Проверяем ассерты для тестирования системы
    print("\n=== Выполнение тестов ===")
    
    # Тест поиска
    assert x.find_computer("server1.misis.ru") is not None, "Компьютер не найден"
    print("✓ Тест поиска пройден")
    
    # Тест независимости копий
    original_server2 = n.find_computer("server2.misis.ru")
    modified_server2 = x.find_computer("server2.misis.ru")
    
    original_components = sum(1 for _ in original_server2.components)
    modified_components = sum(1 for _ in modified_server2.components)
    
    assert original_components == 2, f"Неверное количество компонентов в оригинале: {original_components}"
    assert modified_components == 3, f"Неверное количество компонентов в копии: {modified_components}"
    print("✓ Тест независимости копий пройден")
    
    # Проверка типов дисков
    disk_tests = [
        (Disk(Disk.SSD, 256), "SSD"),
        (Disk(Disk.MAGNETIC, 1000), "HDD")
    ]
    
    for disk, expected_type in disk_tests:
        assert expected_type in str(disk), f"Неверный тип диска в выводе: {str(disk)}"
    print("✓ Тест типов дисков пройден")
    
    print("\nВсе тесты пройдены!")

if __name__ == "__main__":
    main()

File: GitHub_task/task.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cassert>
#include <memory>
#include <sstream>

using namespace std;

class Printable {
public:
    virtual void print(const string& prefix = "", bool isLast = false) const = 0;
    virtual unique_ptr<Printable> clone() const = 0;
    virtual ~Printable() = default;
};

class Component : public Printable {
protected:
    int numeric_val;
public:
    Component(int numeric_val = 0) : numeric_val(numeric_val) {}
};

class Address : public Printable {
    string address;
public:
    Address(const string& addr) : address(addr) {}

    void print(const string& prefix = "", bool isLast = false) const override {
        cout << prefix << (isLast ? "\\-" : "+-") << address << endl;
    }

    unique_ptr<Printable> clone() const override {
        return make_unique<Address>(*this);
    }
};

class CPU : public Component {
    int cores;
    int mhz;
public:
    CPU(int cores, int mhz) : cores(cores), mhz(mhz) {}

    void print(const string& prefix = "", bool isLast = false) const override {
        cout << prefix << (isLast ? "\\-" : "+-") << "CPU, " << cores << " cores @ " << mhz << "MHz" << endl;
    }

    unique_ptr<Printable> clone() const override {
        return make_unique<CPU>(*this);
    }
};

class Memory : public Component {
public:
    Memory(int size) : Component(size) {}

    void print(const string& prefix = "", bool isLast = false) const override {
        cout << prefix << (isLast ? "\\-" : "+-") << "Memory, " << numeric_val << " MiB" << endl;
    }

    unique_ptr<Printable> clone() const override {
        return make_unique<Memory>(*this);
    }
};

class Disk : public Component {
public:
    enum Type { SSD, MAGNETIC };
private:
    Type storage_type;
    vector<pair<int, string>> partitions;
public:
    Disk(Type type, int size) : Component(size), storage_type(type) {}

    Disk& addPartition(int size, const string& name) {
        partitions.emplace_back(size, name);
        return *this;
    }

    void print(const string& prefix = "", bool isLast = false) const override {
        cout << prefix << (isLast ? "\\-" : "+-") << (storage_type == SSD ? "SSD" : "HDD") << ", " << numeric_val << " GiB" << endl;
        string child_prefix = prefix + (isLast ? "  " : "| ");
        for (size_t i = 0; i < partitions.size(); ++i) {
            bool last = (i == partitions.size() - 1);
            cout << child_prefix << (last ? "\\-" : "+-") << "[" << i << "]: " << partitions[i].first << " GiB, " << partitions[i].second << endl;
        }
    }

    unique_ptr<Printable> clone() const override {
        auto copy = make_unique<Disk>(storage_type, numeric_val);
        copy->partitions = partitions;
        return copy;
    }
};

class Computer : public Printable {
    string name;
public:
    vector<unique_ptr<Printable>> addresses;
    vector<unique_ptr<Printable>> components;

    Computer(const string& name) : name(name) {}

    Computer& addAddress(const string& addr) {
        addresses.emplace_back(make_unique<Address>(addr));
        return *this;
    }

    Computer& addComponent(unique_ptr<Printable> comp) {
        components.emplace_back(std::move(comp));
        return *this;
    }

    void print(const string& prefix = "", bool isLast = false) const override {
        cout << prefix << (isLast ? "\\-" : "+-") << "Host: " << name << endl;
        string child_prefix = prefix + (isLast ? "  " : "| ");
        vector<const Printable*> all;
        for (const auto& a : addresses) all.push_back(a.get());
        for (const auto& c : components) all.push_back(c.get());
        for (size_t i = 0; i < all.size(); ++i) {
            all[i]->print(child_prefix, i == all.size() - 1);
        }
    }

    unique_ptr<Printable> clone() const override {
        auto comp = make_unique<Computer>(name);
        for (const auto& a : addresses) comp->addresses.emplace_back(a->clone());
        for (const auto& c : components) comp->components.emplace_back(c->clone());
        return comp;
    }

    const string& getName() const { return name; }

    Computer* find(const string& target_name) {
        if (name == target_name) return this;
        return nullptr;
    }
};

class Network : public Printable {
    string name;
    vector<unique_ptr<Computer>> computers;
public:
    Network(const string& name) : name(name) {}

    Network& addComputer(unique_ptr<Computer> comp) {
        computers.emplace_back(std::move(comp));
        return *this;
    }

    Computer* findComputer(const string& target_name) {
        for (auto& c : computers) {
            if (auto res = c->find(target_name)) return res;
        }
        return nullptr;
    }

    void print(const string& prefix = "", bool isLast = false) const override {
        cout << "Network: " << name << endl;
        for (size_t i = 0; i < computers.size(); ++i) {
            computers[i]->print("", i == computers.size() - 1);
        }
    }

    unique_ptr<Printable> clone() const override {
        auto net = make_unique<Network>(name);
        for (const auto& c : computers) {
            net->addComputer(unique_ptr<Computer>(dynamic_cast<Computer*>(c->clone().release())));
        }
        return net;
    }
};

string capturePrint(const Printable& obj) {
    stringstream buffer;
    streambuf* oldCout = cout.rdbuf(buffer.rdbuf());
    obj.print();
    cout.rdbuf(oldCout);
    return buffer.str();
}

int main() {
    Network n("MISIS network");

    auto server1 = make_unique<Computer>("server1.misis.ru");
    server1->addAddress("192.168.1.1")
           .addComponent(make_unique<CPU>(4, 2500))
           .addComponent(make_unique<Memory>(16000));
    n.addComputer(std::move(server1));

    auto disk = make_unique<Disk>(Disk::MAGNETIC, 2000);
    disk->addPartition(500, "system").addPartition(1500, "data");

    auto server2 = make_unique<Computer>("server2.misis.ru");
    server2->addAddress("10.0.0.1")
           .addComponent(make_unique<CPU>(8, 3200))
           .addComponent(std::move(disk));
    n.addComputer(std::move(server2));

    cout << "=== Созданная сеть ===" << endl;
    n.print();

    string expected_output =
R"(Network: MISIS network
+-Host: server1.misis.ru
| +-192.168.1.1
| +-CPU, 4 cores @ 2500MHz
| \-Memory, 16000 MiB
\-Host: server2.misis.ru
  +-10.0.0.1
  +-CPU, 8 cores @ 3200MHz
  \-HDD, 2000 GiB
    +-[0]: 500 GiB, system
    \-[1]: 1500 GiB, data
)";

    string actual_output = capturePrint(n);
    assert(actual_output == expected_output && "Формат вывода не соответствует ожидаемому");
    cout << "✓ Тест формата вывода пройден" << endl;

    cout << "\n=== Тестирование глубокого копирования ===" << endl;
    auto x = n.clone();
    auto* cloned_network = dynamic_cast<Network*>(x.get());

    cout << "Поиск компьютера server2.misis.ru:" << endl;
    auto* c = cloned_network->findComputer("server2.misis.ru");
    assert(c != nullptr && "Компьютер server2.misis.ru не найден");
    cout << "✓ Найден" << endl;

    cout << "\nДобавляем SSD к найденному компьютеру в копии:" << endl;
    auto new_disk = make_unique<Disk>(Disk::SSD, 500);
    new_disk->addPartition(500, "fast_storage");
    c->addComponent(std::move(new_disk));

    cout << "\n=== Модифицированная копия ===" << endl;
    cloned_network->print();

    cout << "\n=== Исходная сеть (должна остаться неизменной) ===" << endl;
    n.print();

    cout << "\n=== Выполнение тестов ===" << endl;

    auto* server1_clone = cloned_network->findComputer("server1.misis.ru");
    assert(server1_clone != nullptr && "Компьютер server1.misis.ru не найден");
    cout << "✓ Тест поиска пройден" << endl;

    auto* original_server2 = n.findComputer("server2.misis.ru");
    auto* modified_server2 = cloned_network->findComputer("server2.misis.ru");
    int original_components = original_server2->components.size();
    int modified_components = modified_server2->components.size();

    assert(original_components == 2 && "Неверное количество компонентов в оригинале");
    assert(modified_components == 3 && "Неверное количество компонентов в копии");
    cout << "✓ Тест независимости копий пройден" << endl;

    auto disk1 = make_unique<Disk>(Disk::SSD, 256);
    auto disk2 = make_unique<Disk>(Disk::MAGNETIC, 1000);
    string disk1_out = capturePrint(*disk1);
    string disk2_out = capturePrint(*disk2);
    assert(disk1_out.find("SSD") != string::npos && "Неверный тип диска в выводе");
    assert(disk2_out.find("HDD") != string::npos && "Неверный тип диска в выводе");
    cout << "✓ Тест типов дисков пройден" << endl;

    cout << "\nВсе тесты пройдены!" << endl;

    return 0;
}
